### PR TITLE
TP2000-743 Edit multiple measures start and end dates

### PR DESCRIPTION
**Pull request by eadpearce**
# TP2000-743 Edit multiple measures start and end dates

## Why

* Users want to be able to edit start dates as well as end dates on measures

## What

* Adds the end date form and a new start date form to a form wizard
* Users can select which fields they wish to edit at the start of the form

FireShot Capture 050 - Edit measures - Manage Trade Tariffs - localhost
Screenshot 2023-02-17 at 13 48 19
Screenshot 2023-02-17 at 13 48 32

### CODE REVIEW

File: measures/views.py [hunk 1]
```diff
@@ -70,7 +72,31 @@ def get_queryset(self):
...
+    def measure_selections(self):
+        return [
+            SelectableObjectsForm.object_id_from_field_name(name)
+            for name in [*self.session_store.data]
+        ]
+
+
+class MeasureSelectionQuerysetMixin(MeasureSelectionMixin):
+    def get_queryset(self):
+        """Get the measures that are candidates for deletion."""
```
Comment (a-gleeson) [inline measures/views.py:95]: Just for deletion or is it selected candiates?

File: measures/forms.py [hunk 2]
```diff
@@ -1182,3 +1182,76 @@ def clean(self):
                     )
 
         return cleaned_data
+
+
+class MeasureStartDateForm(forms.Form):
```
Comment (a-gleeson) [inline measures/forms.py:1187]: Could the edit fields form be abstracted and reused? Or is it not worth the effort?

File: measures/views.py [hunk 3]
```diff
@@ -163,6 +178,88 @@ def get_context_data(self, **kwargs: Any):
...
+        MeasureEditSteps.END_DATE: {
+            "title": "Edit the end date",
+            "link_text": "End date",
+        },
+    }
+
+    def get_template_names(self):
+        return self.templates.get(
+            self.steps.current,
+            "measures/edit-wizard-step.jinja",
```
Comment (a-gleeson) [inline measures/views.py:218]: Should the default be the start/landing page for edits?

**Comment by a-gleeson** [comment]
Do you think having the form start labeled as just "Start" is clear? When I started reviewing I got a little tripped up between start and start date. Maybe something like "landing" is good although that still may not be clear

File: measures/views.py (hunk 1 above)
Comment (eadpearce) [reply measures/views.py:95]: oop sorry this is a copypaste error. it's selection/editing

File: measures/forms.py (hunk 2 above)
Comment (eadpearce) [reply measures/forms.py:1187]: Yeah not worth it I think. The forms are going to be very slightly different

File: measures/views.py (hunk 3 above)
Comment (eadpearce) [reply measures/views.py:218]: Doesn't really matter. This is only a fallback in case it can't find the given key in the dict

**Comment by eadpearce** [comment]
@a-gleeson Yeah that's a good point. I can change it to something like MeasuresEditFieldsForm ?

**Comment by a-gleeson** [comment]
> @a-gleeson Yeah that's a good point. I can change it to something like MeasuresEditFieldsForm ?

Sounds good to me :)

File: measures/views.py [hunk 4]
```diff
@@ -163,6 +178,88 @@ def get_context_data(self, **kwargs: Any):
         return context
 
 
+@method_decorator(require_current_workbasket, name="dispatch")
+class MeasureEditWizard(
+    MeasureSelectionQuerysetMixin,
+    NamedUrlSessionWizardView,
+):
```
Comment (paulpepper-trade) [inline measures/views.py:185]: I think we need a docstring on this class to explain what it does and how it works. It's an important addition to the way that we manage measures through the UI and we should do our best to minimise the chances of running into those difficulties that we've encountered recently around this area.

**Review by paulpepper-trade — CHANGES_REQUESTED** [review]

File: measures/views.py (hunk 4 above)
Comment (paulpepper-trade) [reply measures/views.py:185]: A docstring would also help us in reviewing this PR.

Comment (eadpearce) [reply measures/views.py:185]: Done

**Review by paulpepper-trade — APPROVED** [review]
I think we'd benefit from more docs about how the wizard is used in this specific case, but it appears we may have a separate activity for that now.

@a-gleeson has given some good comments, so worth getting his approval too?

**Review by a-gleeson — APPROVED** [review]

**Comment by eadpearce** [comment]
@paulpepper-trade Not sure why since everything I've done with the wizard is covered in the django form-tools docs

**Pull request merged**